storage/jellyfish_merkle/src/restore/mod.rs

```diff
@@ -0,0 +1,411 @@
+                            .last_mut()
+                            .expect("This node must exist.")
+                            .set_child(
+                                new_child_index as usize,
+                                ChildInfo::Leaf {
+                                    node: LeafNode::new(new_key, new_value),
+                                },
+                            );
+
+                        break;
+                    }
+                }
+                None => {
+                    // This means that we are going to put a leaf in this position. For all the
+                    // descendants on the left, they are now frozen.
+                    self.freeze(i + 1);
+
+                    // Mark this position as a leaf child.
+                    self.partial_nodes[i].set_child(
+                        child_index,
+                        ChildInfo::Leaf {
+                            node: LeafNode::new(new_key, new_value),
+                        },
+                    );
+
+                    // We do not add this leaf node to self.frozen_nodes because we don't know its
+                    // node key yet. We will know its node key when the next account comes.
+                    break;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Puts the nodes that will not be changed in `self.frozen_nodes`.
+    fn freeze(&mut self, target_len: usize) {
+        // Freeze the rightmost leaf node on the lowest level. This node was inserted in the
+        // previous `restore_one` call.
+        let last_node = self
+            .partial_nodes
+            .last()
+            .expect("Must have at least one partial node.");
+        for i in (0..16).rev() {
+            if let Some(ref child_info) = last_node.children[i] {
+                if let ChildInfo::Leaf { node } = child_info {
+                    let child_node_key = last_node
+                        .node_key
+                        .gen_child_node_key(self.version, (i as u8).into());
+                    self.frozen_nodes
+                        .insert(child_node_key, node.clone().into());
+                    break;
+                }
+            }
+        }
+
+        // Freeze extra internal nodes.
+        while self.partial_nodes.len() > target_len {
+            let last_node = self.partial_nodes.pop().expect("This node must exist.");
+            let (node_key, internal_node) = last_node.into_internal_node(self.version);
+            let node_hash = internal_node.hash();
+            self.frozen_nodes.insert(node_key, internal_node.into());
+
+            // Now that we have computed the hash of the internal node above, we will also update
+            // its parent unless it is root node.
+            if let Some(parent_node) = self.partial_nodes.last_mut() {
+                // This internal node must be the rightmost child of its parent at the moment.
+                for i in (0..16).rev() {
```

Comment:
Can you do the same `rposition(|x| x.is_some())` as above?  It saves having to change 0..16 all over the place should we ever change the size of the array.